src/EditorFeatures/Core/Implementation/RenameTracking/RenameTrackingDiagnosticAnalyzer.cs

```diff
@@ -7,7 +7,7 @@
 namespace Microsoft.CodeAnalysis.Editor.Implementation.RenameTracking
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
-    internal sealed class RenameTrackingDiagnosticAnalyzer : DiagnosticAnalyzer, IBuiltInAnalyzer
+    internal sealed class RenameTrackingDiagnosticAnalyzer : DiagnosticAnalyzer, IBuiltInAnalyzer, IInProcessAnalyzer
```

Comment:
while it's fairly obvious, it would probably be good to document why any particular analyzer is marked as IInProc.

Reply:
In all cases, they are marked as in-process because they were implicitly restricted to in-process prior to this change. I did not attempt to determine _why_ any individual analyzer is marked as in-process.

Reply:
so, we had IBuiltInAnalyzer.RunInProc and I had to remove it since I got code review saying that is bad, can't understand, too complex or whatever. and changed to OpenFileOnly since somehow that's better. 

https://github.com/dotnet/roslyn/commit/0999ff02597f2e9f17bb2b9b4f598fdd1e502e65#diff-8631138eee12337d6b8b5de95567e3d4

so, now, It is changed back to same thing but as interface. so, wondering why we need it?

…

OOP doesn't have concept of open or closed file. so any out of proc analyzer marked as OpenFileOnly doesn't make sense. so, having 2 flags, I am not sure what that means.

Reply:
> so, now, It is changed back to same thing but as interface. so, wondering why we need it?

My goal is to remove both of these concepts. Separating them helps with reviewing to determine where/why each flag is needed.

> OOP doesn't have concept of open or closed file. so any out of proc analyzer marked as OpenFileOnly doesn't make sense. so, having 2 flags, I am not sure what that means.

The primary case where this is relevant is Fix All operations. OpenFileOnly no longer forces in-process analysis for analyzers that are equally fine running out-of-process. However, keep in mind that the long-term goal is eliminating the need for both of these flags.

Reply:
> The primary case where this is relevant is Fix All operations, which force full-project (or solution) analysis for analyzers even when they are marked OpenFileOnly. OpenFileOnly no longer forces in-process analysis for analyzers that are equally fine running out-of-process. However, keep in mind that the long-term goal is eliminating the need for both of these flags.

I see. so, what you are saying is all OpenFileOnly analyzer should be able to run out of proc? and for fix all, you need to run analyzers out of proc? 

if so, this condition is wrong.
var inProcResultTask = AnalyzeInProcAsync(CreateAnalyzerDriver(analyzerDriver, a => (forcedAnalysis || !a.IsOpenFileOnly(workspace)) && a.IsInProcessOnly()), project, remoteHostClient, cancellationToken);

you need to change it to
(!forcedAnalysis && !a.IsOpenFileOnly(workspace)) && a.IsInProcessOnly())

Reply:
actually, are you sure that all OpenFileOnly analyzer can run out of proc? or are you planning to make all analyzer runnable out of proc? 

anyway, if you make them happen (or if it can already does that), you can just remove this line actually.
var inProcResultTask = AnalyzeInProcAsync(CreateAnalyzerDriver(analyzerDriver, a => (forcedAnalysis || !a.IsOpenFileOnly(workspace)) && a.IsInProcessOnly()), project, remoteHostClient, cancellationToken);

since this code path only get called when "fix all" or "full solution analysis" is on.

Reply:
> actually, are you sure that all OpenFileOnly analyzer can run out of proc?

I marked all analyzers that previously returned `true` as in-process only, even in cases where I believe they would still work if they were out of process (e.g. naming styles analyzer). Some analyzers *conditionally* returned true for OpenFileOnly; I assumed that since they only conditionally returned true then there was no inherent need to run them in-process. These analyzers are not marked as in-process only (e.g. simplify type name).

> anyway, if you make them happen (or if it can already does that), you can just remove this line actually.

I plan to do that in a follow-up PR to remove the `IInProcessAnalyzer` interface.

Reply:
> (forcedAnalysis || !a.IsOpenFileOnly(workspace)) && a.IsInProcessOnly()

Here is my thinking on the current condition:

* Only in-process analyzers will run in-process, hence `&& a.IsInProcessOnly()`
* For `forcedAnalysis`
    * If `forcedAnalysis` is true, we want to run all analyzers
    * If `forcedAnalysis` is false, we only want to run analyzers that are not marked OpenFileOnly

Reply:
> I assumed that since they only conditionally returned true then there was no inherent need to run them in-process. These analyzers are not marked as in-process only (e.g. simplify type name).

I am fine with playing with these. just to make sure that we don't have live analysis with full solution analysis on doesn't make VS unusable.

Reply:
here is the existing behavior.

1. we never run analyzers in proc for closed file except "fix all" case where user explicitly asked to run.
2. we never run OpenFileOnly analyzers in OOP since there is no such thing as Open file in OOP
3. for fix all case, we run OpenFileOnly analyzers in proc for full solution.